small/region_asan.c

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "region.h"
+
+/** Allocate new memory block whether for allocation or reservation. */
+static SMALL_NO_SANITIZE_ADDRESS void *
+region_prepare_buf(struct region *region, size_t size, size_t alignment,
+		   size_t used)
+{
+	struct region_allocation *alloc =
+			small_asan_alloc(size, alignment,
+					 sizeof(struct region_allocation));
+	alloc->used = used;
+	alloc->alignment = alignment;
+	rlist_add_entry(&region->allocations, alloc, link);
+
+	return small_asan_payload_from_header(alloc);
+}
+
+void *
+region_aligned_reserve(struct region *region, size_t size, size_t alignment)
+{
+	small_asan_assert(region->reserved == 0);
+
+	/* See explanation for lower limit in the header. */
+	size_t pagesize = small_getpagesize();
```

Comment:
Come to think of it, I don't see any point in using page size here. Let's just hard-code the constant.

Reply:
Discussed f2f to have much lower limit to reduce memory waste.
```diff
@@ -50,10 +50,10 @@ region_aligned_reserve(struct region *region, size_t size, size_t>
 {
        small_asan_assert(region->reserved == 0);

-       /* See explanation for lower limit in the header. */
-       size_t pagesize = small_getpagesize();
-       if (size < pagesize)
-               size = pagesize;
+       /* See limit explanation in the header. */
+       int lower_limit = 128;
+       if (size < lower_limit)
+               size = lower_limit;
```
```diff
- * Also minimum reserved size equals to pagesize. This allows to follow
- * similar code paths as in case of regular implementation at least in
- * interesting cases.
+ * Regular implementation reserves memory in blocks (slabs which size grows
+ * exponentially). We try to imitate this by reserving at least 128 bytes
+ * so that runtime code path with ASAN implementation is similar to regular one.
+ * The overreserve is not very large comparing to regular one to reduce memory
+ * waste.
```